hyperstart/multicast.go

```diff
@@ -0,0 +1,158 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package hyperstart
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	hyper "github.com/hyperhq/runv/hyperstart/api/json"
+)
+
+type ctlDataType string
+
+const (
+	eventType ctlDataType = "ctlEvent"
+	replyType ctlDataType = "ctlReply"
+)
+
+type multicast struct {
+	reply []chan *hyper.DecodedMessage
+	event map[string]chan *hyper.DecodedMessage
+	ctl   net.Conn
+	mutex sync.Mutex
```

Comment:
A nice pattern is to use struct embedding here (just sync.Mutex) so the locking becomes multicast.Lock()